src/main/scala/is/hail/methods/MaximalIndependentSet.scala

```diff
@@ -0,0 +1,101 @@
+package is.hail.methods
+
+import org.apache.spark.SparkContext
+import org.apache.spark.rdd.RDD
+import org.apache.spark.graphx._
+
+import scala.reflect.ClassTag
+
+object MaximalIndependentSet {
+
+  def apply[VD: ClassTag, ED: ClassTag](g: Graph[VD, ED]): Set[Long] = {
+    // Initially set each vertex to its own degree.
+    // Start a pregel run, everyone passing a (ID, degree) pair.
+    //  -On message reception, if current degree is greater than received degree, update status to to reflect this, alert everyone
+    //  -If current degree <= received degree, status does not change, don't bother sending alert.
+
+    //Every vertex needs to know
+    // -Whether it changed recently (some facet of old state must be stored)
+    // -What the highest degree message received has been
+    // -What vertex is in that highest degree message
+
+    var graph = Graph[((VertexId, Int), VertexId), ED](g.collectNeighborIds(EdgeDirection.Either)
+      .map{ case(v, neighbors) => (v, ((v, neighbors.size), -1L))}, g.edges: RDD[Edge[ED]])
+
+    val initialMsg = (-1L, -1)
+
+    //vertexId is ID of the receiver
+    def vprog(vertexId: VertexId, value: ((VertexId, Int), VertexId), message: (VertexId, Int)): ((VertexId, Int), VertexId) = {
+      val ((currentVertex, currentMaxDegree), oldVertex) = value
+      val (receivedVertex, receivedMaxDegree) = message
+
+
+      if (message == initialMsg) {
+        value
+      }
+      else if (currentMaxDegree < receivedMaxDegree) {
+        (message, currentVertex)
+      }
+      else {
+        (value._1, currentVertex)
+      }
+
+    }
+
+    def sendMsg(triplet: EdgeTriplet[((VertexId, Int), VertexId), ED]): Iterator[(VertexId, (VertexId, Int))] = {
+      val sourceAttr = triplet.srcAttr
+
+      if (sourceAttr._1._1 == sourceAttr._2) {
+        Iterator.empty
+      }
+      else {
+        Iterator((triplet.dstId, sourceAttr._1))
+      }
+    }
+
+    def mergeMsg(msg1: (VertexId, Int), msg2: (VertexId, Int)): (VertexId, Int) = {
+      if (msg1._2 > msg2._2) msg1 else msg2
+    }
+
+    var atLeastOneEdge = graph.numEdges > 0
+    while(atLeastOneEdge) {
+      graph = graph.pregel(initialMsg, Int.MaxValue, EdgeDirection.Both)(vprog, sendMsg, mergeMsg)
+      graph = graph.subgraph(_ => true, (id, value) => value._1._1 != value._2 || value._1._2 == 0)
+
+      if (graph.numEdges == 0) {
+        atLeastOneEdge = false
+      }
+    }
```

Comment:
Yes, thank you. When I was writing it I had a quick moment of doubt as to whether the "graph" identifier in the loop condition would be evaluated once and constantly refer to same object from when loop started or would change as the variable changed.